contracts/dynamic-contracts/ExtensionRegistry.sol

```diff
@@ -34,9 +36,31 @@ contract ExtensionRegistry is IExtensionRegistry, ExtensionState, PermissionsEnu
         _updateExtension(_extension);
     }
 
-    /// @notice Remove an existing extension from the registry.
-    function removeExtension(string memory _extensionName) external onlyRole(DEFAULT_ADMIN_ROLE) {
-        _removeExtension(_extensionName);
+    /// @notice Adds an extension to an extension snapshot.
+    function buildExtensionSnapshot(
+        string memory _extensionSnapshotId,
+        string[] memory _extensionNames,
+        bool _freeze
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(bytes(_extensionSnapshotId).length > 0, "ExtensionRegistry: extension snapshot ID cannot be empty.");
+
+        uint256 len = _extensionNames.length;
+        for (uint256 i = 0; i < len; i += 1) {
+            _addExtensionToSnapshot(_extensionSnapshotId, _extensionNames[i]);
+        }
+
+        if (_freeze) {
+            _freezeExtensionSnapshot(_extensionSnapshotId);
+        }
+
+        emit ExtensionSnapshotUpdated(_extensionSnapshotId, _extensionNames);
+    }
+
+    /// @dev Registers a router contract with an extension snapshot as its default set of extensions.
+    function registerWithSnapshot(string memory _extensionSnapshotId) external {
```

Comment:
anyone can call this function? that seems problematic no?

what if we check a `supportsInterface` call on the msg.sender here?

Reply:
Yes I'll add that. It won't ultimately affect how the contract's working, but it'll be a good practice to use it.

Reply:
Update on this: we cannot add a supports interface check in the function since the function is called inside the constructor of the router contract, and the following fails for a contract being constructed (since no `code` exists at the address yet):

```solidity
if(router.code.length > 0) {
  require(ERC165(router).supportsInterface(type(IRouter).interfaceId), "Error");
}
```